src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx

```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
     usePortal: false
   };
 
+  const searchDrawerProps = {
+    onClose: handleCloseSearch,
+    autoFocus: true,
+    canEscapeKeyClose: true,
+    canOutsideClickClose: true,
+    enforceFocus: true,
+    hasBackdrop: true,
+    isOpen: isSearchOpen,
+    position: Position.LEFT,
+    usePortal: false
+  };
+
+  function queryIndexTrie(query: string) {
+    let node = indexTrie;
+    for (let i = 0; i < query.length; i++) {
+      const char = query[i];
+      if (node[char]) {
+        node = node[char];
+      } else {
+        return [];
+      }
+    }
+    if (node['value']) {
+      return node['value'];
+    }
+    return [];
+  }
+
+  function queryTextTrie(query: string) {
+    let node = textTrie;
+    for (let i = 0; i < query.length; i++) {
+      const char = query[i];
+      if (node[char]) {
+        node = node[char];
+      } else {
+        return [];
+      }
+    }
+    if (node['value']) {
+      return node['value'];
+    }
+    return [];
+  }
+
+  function autoComplete(str: string, limit: number, jsonData: any) {
+    if (str.length === 0) {
+      return [];
+    }
+    str = str.toLowerCase();
+    const trie = jsonData;
+
+    function next(node: any, i: number) {
+      if (i === 26) return node[' '];
+      else if (i >= 0 && i <= 25) {
+        return node[String.fromCharCode(i + 97)];
+      }
+      return null;
+    }
+    const letters = [
+      'a',
+      'b',
+      'c',
+      'd',
+      'e',
+      'f',
+      'g',
+      'h',
+      'i',
+      'j',
+      'k',
+      'l',
+      'm',
+      'n',
+      'o',
+      'p',
+      'q',
+      'r',
+      's',
+      't',
+      'u',
+      'v',
+      'w',
+      'x',
+      'y',
+      'z',
+      ' '
+    ];
+
+    function toEnd(query: any) {
+      let node = trie;
+      for (let i = 0; i < query.length; i++) {
+        const char = query[i];
+        if (node[char]) {
+          node = node[char];
+        } else {
+          return null;
+        }
+      }
+      return node;
+    }
+
+    function recur(node: any, path: string) {
+      if (node == null || limit < 0) {
+        return;
+      }
+      if (node.value) {
+        ans.push(path);
+        limit--;
+      }
+      for (let i = 0; i < 27; i++) {
```

Comment:
What is 27? Is this a magic number?

Reply:
Similarly, what does the `voidSearch` function do? The readability is slightly hampered by the functions being called out-of-order they're defined in.

Reply:
In that case, what does `queryTrie` do?

Reply:
for index search, since all the index are in the IndexTrie, we can just use queryTire index;

But for user search, since the textTrie only contains the occurrences of individual words rather than sentences, for example, "instruction", but not "follow instructions", I need to use voidSearch instead